Implement getItemURL for Dropbox

Ragnis opened the pull request:
#1052

Commit:
Make BaseClient#getItemURL async

Commit:
Re-implement Dropbox#getItemURL

skddc commented:
@Ragnis Is this actually still work in progress? Looks to me like this is basically done.

skddc commented:
@Ragnis Would be cool if you could update us on the status here. See my question from November.

Ragnis committed:
Remove an old test

The test assumes that if a file is successfully fetched from a /public/
directory then the download URL can also be used as a public URL. That
is however not true since DropBox requires all public URLs to be
explicitly created.

Ragnis commented:
I guess it is done. I re-based the code to latest master and also fixed the tests.

raucao reviewed (changes requested):
I just played around with this branch in Sharesome and found a logical bug with the implementation:

The way the function works on a normal BaseClient is that the path argument is relative to the client scope. So when used in a module, the `publicClient` would have a base of `mymodulename/public/`. Hence the `makePath()` call here: https://github.com/remotestorage/remotestorage.js/blob/master/src/baseclient.js#L299

However, this implementation doesn't add the base path, and so `_itemRefs` aren't matched correctly when calling `getItemURL()` with a relative path (in the shares module that would just be a filename e.g.). This results in `undefined` being returned instead of the stored public URL from `_itemRefs`.

---

Another thing, but I'm not 100% sure how to solve it, is that `storeFile()` actually returns a URL on a normal BaseClient, but we only overwrite `put` in the Dropbox class, and returning a simple URL from that broke it for me when I tried to add it.

raucao commented:
Closing in favor of #1389
